### PR TITLE
WIP: Calendar

### PR DESCRIPTION
**Pull request by chrisadubois**
Once review is complete I will squash the commits, and add tests.

we will need the calendaring to work for interstitial page, we have to get the calendar events for a person, to both populate the calendar in the web client for 1 day before and 1 week ahead.

we will need the calendar events to send events to the client for updates to the schedule

### CODE REVIEW

File: packages/node_modules/@webex/internal-plugin-calendar/src/calendar.js [hunk 1]
```diff
@@ -4,9 +4,70 @@
...
+  },
+
+  _handleDelete(data) {
+    const item = CalendarCollection.remove(data.calendarMeetingExternal.id);
+
+    this.trigger('calendar:meeting:delete', item);
+  },
+
+  // default to yesterday to end of the week so 8 days total
+  populate() {
```
Comment (andrewholsted) [inline packages/node_modules/@webex/internal-plugin-calendar/src/calendar.js]: What do yo think about allowing properties to be passed in to the method to override the default config? If that's too much flexibility that's fine

Comment (andrewholsted) [reply packages/node_modules/@webex/internal-plugin-calendar/src/calendar.js]: do you need to return this.list()?

File: packages/node_modules/@webex/internal-plugin-calendar/src/util.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+const CalendarUtil = {
+
+  derive(item) {
+    item.endTime = new Date(new Date(item.start).getTime() + item.durationMinutes * 60000);
+
+    return item;
```
Comment (andrewholsted) [inline packages/node_modules/@webex/internal-plugin-calendar/src/util.js]: Some code comments would go a long way here. I don't know the purpose/intent of this.

File: packages/node_modules/@webex/internal-plugin-calendar/src/calendar.js (hunk 1 above)
Comment (chrisadubois) [reply packages/node_modules/@webex/internal-plugin-calendar/src/calendar.js]: this should be done at startup time, and we re format to an object, unnecessary to return

i think I'm cool with properties

File: packages/node_modules/@webex/internal-plugin-calendar/src/util.js (hunk 2 above)
Comment (chrisadubois) [reply packages/node_modules/@webex/internal-plugin-calendar/src/util.js]: will add

File: packages/node_modules/@webex/internal-plugin-calendar/src/calendar.js [hunk 3]
```diff
@@ -4,9 +4,70 @@
 
 import {WebexPlugin} from '@webex/webex-core';
 
+import CalendarCollection from './collection';
+
 const Calendar = WebexPlugin.extend({
   namespace: 'Calendar',
 
+  register() {
```
Comment (adamweeks) [inline packages/node_modules/@webex/internal-plugin-calendar/src/calendar.js]: Could you add device.refresh and mercury.connect here and move the listen events to a `listenToEvents` function like the meetings plugin?

Comment (chrisadubois) [reply packages/node_modules/@webex/internal-plugin-calendar/src/calendar.js]: yes

File: packages/node_modules/@webex/internal-plugin-calendar/src/calendar.js [hunk 4]
```diff
@@ -2,10 +2,225 @@
...
+     */
+  register() {
+    if (!this.webex.canAuthorize) {
+      this.logger.error('calendar->register#ERROR, Unable to register, SDK cannot authorize');
+
+      return Promise.reject(new Error('SDK cannot authorize'));
+    }
+
+    if (this.registered) {
+      this.logger.info('calendar->register#INFO, Meetings plugin already registered');
```
Comment (adamweeks) [inline packages/node_modules/@webex/internal-plugin-calendar/src/calendar.js]:
```suggestion
      this.logger.info('calendar->register#INFO, Calendar plugin already registered');
```

File: packages/node_modules/@webex/internal-plugin-calendar/src/calendar.js [hunk 5]
```diff
@@ -2,10 +2,225 @@
...
+     * Explicitly tears down the calendar plugin by deregistering
+     * the device, disconnecting from mercury, and stops listening to calendar events
+     *
+     * @returns {Promise}
+     * @public
+     * @memberof Meetings
+     */
+  unregister() {
+    if (!this.registered) {
+      this.logger.info('meetings->unregister#INFO, Meetings plugin already unregistered');
```
Comment (adamweeks) [inline packages/node_modules/@webex/internal-plugin-calendar/src/calendar.js]:
```suggestion
      this.logger.info('calendar->unregister#INFO, Calendar plugin already unregistered');
```

File: packages/node_modules/@webex/internal-plugin-calendar/src/calendar.js [hunk 6]
```diff
@@ -2,10 +2,225 @@
...
+      });
+  },
+
+  /**
+   * registers for calendar events through mercury
+   * @returns {undefined}
+   */
+  listenForEvents() {
+    // Calendar mercury events listener
+    this.webex.internal.mercury.on('event:calendar.meeting.create', (envelope) => {
```
Comment (adamweeks) [inline packages/node_modules/@webex/internal-plugin-calendar/src/calendar.js]: Can we make these event string constants like we have them in the meetings plugin as `CALENDAR_EVENTS`?

File: packages/node_modules/@webex/internal-plugin-calendar/src/calendar.js [hunk 7]
```diff
@@ -2,10 +2,225 @@
...
+      return CalendarCollection.getAll();
+    });
+  },
+
+  /**
+   * bucket meetings by date
+   * @todo
+   * @returns {undefined}
+   */
+  sortByDate() {
```
Comment (adamweeks) [inline packages/node_modules/@webex/internal-plugin-calendar/src/calendar.js]: Can we remove this un-implemented function?

File: packages/node_modules/@webex/internal-plugin-calendar/src/calendar.js [hunk 8]
```diff
@@ -2,10 +2,225 @@
...
+    // TODO:
+  },
+
+  /**
+   * get the calendar item that has a matching value
+   * @param {String} key meeting property
+   * @param {Any} value the meeting property to match
```
Comment (adamweeks) [inline packages/node_modules/@webex/internal-plugin-calendar/src/calendar.js]:
```suggestion
   * @param {String} value the meeting property to match
```

**Comment by adamweeks** [comment]
There seems to be meetings plugin changes in this PR from @arun3528, is that supposed to be here?

**Comment by adamweeks** [comment]
I'm closing this as stale. No updates have happened and comments have not been addressed. Please feel free to reopen when are you are ready for our team to review.

**Pull request closed without merging**